chain/network/src/tests/cache_edges.rs

```diff
@@ -1,311 +1,221 @@
-        on_disk_peers: Vec<(usize, u64)>,
-    ) {
-        let on_memory = on_memory.into_iter().map(EdgeDescription::from).collect::<HashSet<_>>();
-        let on_disk_edges = on_disk_edges
-            .into_iter()
-            .map(|(key, value)| {
-                (
-                    key,
-                    value
-                        .into_iter()
-                        .map(|(u, v, t)| {
-                            let peer0 = self.get_peer(u).clone();
-                            let peer1 = self.get_peer(v).clone();
-                            let (u, v) = if peer1 < peer0 { (v, u) } else { (u, v) };
-                            EdgeDescription::from((u, v, t))
-                        })
-                        .collect::<HashSet<_>>(),
-                )
-            })
-            .collect::<HashMap<_, _>>();
-        let on_disk_peers = on_disk_peers.into_iter().collect::<HashMap<_, _>>();
-
-        // Check memory edges
-        for EdgeDescription(peer0, peer1, edge_type) in on_memory.iter() {
-            let peer0 = self.get_peer(*peer0).clone();
-            let peer1 = self.get_peer(*peer1).clone();
-            let (peer0, peer1) = Edge::make_key(peer0, peer1);
-
-            let res = self.routing_table.edges_info.get(&(peer0, peer1));
-            assert!(res.is_some());
-            let edge = res.unwrap();
-            assert_eq!(edge.edge_type(), *edge_type);
-        }
-        let active_edges = on_memory
-            .iter()
-            .filter_map(|x| if x.2 == EdgeState::Active { Some(1) } else { None })
-            .count();
-
-        assert_eq!(active_edges, self.routing_table.raw_graph.total_active_edges() as usize);
-        assert_eq!(
-            active_edges,
-            self.routing_table.raw_graph.compute_total_active_edges() as usize
-        );
-        assert_eq!(on_memory.len(), self.routing_table.edges_info.len());
-
-        // Check for peers on disk
-        let mut total_peers = 0;
-        for (peer, nonce) in self.store.iter(DBCol::PeerComponent) {
-            total_peers += 1;
-
-            let peer = PeerId::try_from_slice(peer.as_ref()).unwrap();
-            let nonce = u64::try_from_slice(nonce.as_ref()).unwrap();
-            let peer_ix = self.rev_peers.get(&peer).unwrap();
-            let res = on_disk_peers.get(peer_ix).unwrap();
-            assert_eq!(*res, nonce);
-        }
-        assert_eq!(total_peers, on_disk_peers.len());
-
-        // Check for edges on disk
-        let mut total_nonces = 0;
-        for (nonce, edges) in self.store.iter(DBCol::ComponentEdges) {
-            total_nonces += 1;
-
-            let nonce = u64::try_from_slice(nonce.as_ref()).unwrap();
-
-            let edges = Vec::<Edge>::try_from_slice(edges.as_ref()).unwrap();
-            let current_edges = on_disk_edges.get(&nonce).unwrap();
-
-            assert_eq!(edges.len(), current_edges.len());
+    fn new() -> Self {
+        let mut rng = util::make_rng(87927345);
```

Comment:
> is easy to fix flake (just re-run locally with the same seed)
AFAIK the state-of-the-art is to use fixed seeds, especially to not lose reproducibility for random (pun intended) reasons.
How would you know what the seed was? Imagine someone having a failing test locally and reporting it to us. We then need to actually ask that person to find and provide us with seed that they used. Having a test which is random, means that there might be a flake somewhere, which subconsciously makes people restart the test multiple times before even starting debugging, hoping that they can make it almost always pass (and don't investigate the issue) and pretend they never seen the actual problem (i.e. claim it is not reproducible).

> actually uncovers a bug
Using a random seed is misleading people to think that we have "larger" coverage compared to a fixed seed (the seed space is enormous, using a small bunch of random seeds for tests executed from time to time doesn't fix that issue).
Here we use PRNG only to explicitly indicate that ANY value will do, and we just want to have it non-trivial.
If you want a better, well distributed coverage then you either
- use fuzzing for heuristic search of interesting cases (and I'd use it in addition to having hand-picked set of cases, rather that instead, because I don't trust heuristics to select cases that I find important to cover).
- run the test in a loop (reusing PRNG created once with a fixed seed), and generate the inputs specifically according to that distribution.

Reply:
>AFAIK the state-of-the-art is to use fixed seeds

Yeah, I also got the feeling that this is state of the art, but this is one piece of common knowledge I don't yet understand. 

>Using a random seed is misleading people to think that we have "larger" coverage compared to a fixed seed

So this is something I don't necessary agree with. In my experience, it often happens in practice that a randomized test fails only sometimes (like, 20% of the time). Sure, if you add a top-level loop to every test, that'd fix the issue, but:

* that requires not forgetting that every randomized tests needs a loop
* that increases testing time multiplicatively  

I do think that randomizing seed is a cheap way to best-effort improve coverage. 

The biggest issue I see with fixed seeds is that they might mask incorrect tests. What if the randomized test is buggy and would fail for a different seed? Fixing the seed to me seems to be in the same class as adding/increasing timeouts -- it kinda rids you off the flakes, but for the wrong reason. 


>and pretend they never seen the actual problem (i.e. claim it is not reproducible).

If this fails on CI and CI prints the seed which was used, that's easy to repro and hard to deny. It's inconvenient that it might fail for a different PR, but seems better that not knowing about the problem? 


No super strong opinion here either!

Reply:
Just my 0.02Ⓝ:

> What if the randomized test is buggy and would fail for a different seed? 

You can say the same about any other test.  If you write test with no randomness, i.e. devise a test scenario by hand, that test may fail if you change the scenario.

> Fixing the seed to me seems to be in the same class as adding/increasing timeouts -- it kinda rids you off the flakes, but for the wrong reason.

If you have a test which fails and you fix it by changing the seed than that’s the same as increasing timeouts (or maybe even worse) but if you write a test with a fixed seed then that’s just like writing a ‘assert_eq!("foobar", concat("foo", "bar"));` test.  ‘foo’ and ‘bar’ are just some string you’ve pulled out of a hat just like the random seed is just some number you’ve pulled out of a hat.

> If this fails on CI and CI prints the seed which was used

That’s definitely important if random seed is used; the seed must be printed so the test can be reproduced.

Reply:
> Yeah, I also got the feeling that this is state of the art, but this is one piece of common knowledge I don't yet understand.

The common knowledge here is that non-determinism too often means non-reproducibility and hard-to-debug flakiness.
For small systems/simple functions when there is only one possible spot at which it fails at, let's say, 20% rate it is not that hard to debug. But when you have a function called 100 times in a test in different contexts and it can fail at 0.1% at any of them, then you get 10% repro of failure but for 100 different reasons. It gets even worse if you have a bunch of non-deterministic functions used, when each of them fails at a different very small rate.

> The biggest issue I see with fixed seeds is that they might mask incorrect tests.

as @mina86 says, the test is not incorrect only because it looks at a specific case - this is the property of any test. Tests do not cover the whole input space (usually), and that's why formal proofs are superior to tests ;).

Reply:
>the test is not incorrect only because it looks at a specific case 

Hm, I think this is a specific disagreement! If a randomized test passes with one seed, but fails with a different seed, I would definitely consider that to be a bug in the test and something which should be fixed (as in, if we don't fix this know, we likely would have to fix this later anyway, spending more time on it).

Reply:
Let's say that you have detected a seed on which your test failed. Then you fix the bug in the implementation. What would you do with your randomized test in that case? Would you leave it as is, or would you hard code another test equivalent to the one with the detected problematic seed, so that the specific bug you just fixed won't ever reappear? Would you consider the newly created test buggy if your randomized test detected yet another bug in implementation later?

AFAIC
- a test with different seed is a different test which just tests different input
- if there is a bug in an implementation, it doesn't make succeeding tests broken. It just proves that more tests are needed.
- you never know if your implementation is covered enough with tests, unless you test all inputs. The feeling of sufficient coverage is subjective.
- I like my code being deterministic

Still, I respect your opinion. Not sure if we can reach an agreement on this aspect this time.